server/src/main/java/org/elasticsearch/action/admin/cluster/configuration/TransportAddVotingTombstonesAction.java

```diff
@@ -0,0 +1,178 @@
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Setting.Property;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.threadpool.ThreadPool.Names;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class TransportAddVotingTombstonesAction extends TransportMasterNodeAction<AddVotingTombstonesRequest, AddVotingTombstonesResponse> {
+
+    public static final Setting<Integer> MAXIMUM_VOTING_TOMBSTONES_SETTING
+        = Setting.intSetting("cluster.max_voting_tombstones", 10, 1, Property.Dynamic, Property.NodeScope);
+
+    @Inject
+    public TransportAddVotingTombstonesAction(TransportService transportService, ClusterService clusterService, ThreadPool threadPool,
+                                              ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(AddVotingTombstonesAction.NAME, transportService, clusterService, threadPool, actionFilters, AddVotingTombstonesRequest::new,
+            indexNameExpressionResolver);
+    }
+
+    @Override
+    protected String executor() {
+        return Names.SAME;
+    }
+
+    @Override
+    protected AddVotingTombstonesResponse newResponse() {
+        throw new UnsupportedOperationException("usage of Streamable is to be replaced by Writeable");
+    }
+
+    @Override
+    protected AddVotingTombstonesResponse read(StreamInput in) throws IOException {
+        return new AddVotingTombstonesResponse(in);
+    }
+
+    @Override
+    protected void masterOperation(AddVotingTombstonesRequest request, ClusterState state,
+                                   ActionListener<AddVotingTombstonesResponse> listener) throws Exception {
+
+        clusterService.getMasterService().submitStateUpdateTask("add-voting-tombstones", new ClusterStateUpdateTask() {
+
+            final ClusterStateObserver observer
+                = new ClusterStateObserver(clusterService, request.getTimeout(), logger, threadPool.getThreadContext());
+
+            private Set<DiscoveryNode> resolvedNodes;
+
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                final DiscoveryNodes allNodes = currentState.nodes();
+                assert resolvedNodes == null : resolvedNodes;
+                resolvedNodes = Arrays.stream(allNodes.resolveNodes(request.getNodeDescriptions()))
+                    .map(allNodes::get).filter(DiscoveryNode::isMasterNode).collect(Collectors.toSet());
+
+                if (resolvedNodes.isEmpty()) {
+                    throw new IllegalArgumentException("add voting tombstones request for " + Arrays.asList(request.getNodeDescriptions())
```

Comment:
should we do these already as a pre-flight check before submitting the cluster state update task (and do it here again)? This avoids in each (possibly invalid) user action resulting in a cluster state update task.

Reply:
Ok, done that.